tests/Unit/DataProcessorTest.php

```diff
@@ -0,0 +1,87 @@
+use Piwik\Plugins\CampaignVisitorsByTime\Test\TestDataProcessor;
+
+/**
+ * @covers \Piwik\Plugins\CampaignVisitorsByTime\DataProcessor
+ */
+class DataProcessorTest extends \PHPUnit\Framework\TestCase {
+
+	const TEST_REFERER = 'test_referer';
+	const TEST_HOUR = 5;
+	const TEST_MINUTES = '00';
+	const TEST_MINUTES_ALT = '15';
+
+	public function testWhenMultipleKeywordsAreUsed_campaignValuesAreCalculatedCorrectly() {
+		$rows = $this->newZendDbRowsMock( $this->getMockDatabaseData() );
+		$dataProcessor = new TestDataProcessor();
+		$dataProcessor->processDatasets( $rows );
+		$this->assertEquals(
+			40,
+			$dataProcessor->getCampaigns()[self::TEST_REFERER][$this->getStoredHour(
+				self::TEST_HOUR,
+				self::TEST_MINUTES
+			)],
+			'Campaign total for given time should sum up all keywords for specific time of day with timezone offset'
+		);
+	}
+
+	public function testWhenKeywordsAreStored_visitorNumbersAreCorrectlyAssignedToTime() {
+		$rows = $this->newZendDbRowsMock( $this->getMockDatabaseData() );
+		$dataProcessor = new TestDataProcessor();
+		$dataProcessor->processDatasets( $rows );
+		$keywordData = $dataProcessor->getCampaignKeywordData( self::TEST_REFERER );
+		$this->assertEquals(
+			33,
+			$keywordData['testkeyword04'][$this->getStoredHour( self::TEST_HOUR, self::TEST_MINUTES_ALT )],
+			'Individual keywords store their own value per campaign for a specific given 15 minutes block'
+		);
+	}
+
+	public function newZendDbRowsMock( array $rows ) {
+		$dbRows = $this->createMock( \Zend_Db_Statement::class );
+		$dbRows->method( 'fetch' )->willReturnOnConsecutiveCalls( ...$rows );
+		return $dbRows;
+	}
+
+	private function getHourInTimeZone( $hour ): int {
+		$timezone = new \DateTimeZone( 'Europe/Berlin' );
+		return $hour + $timezone->getOffset( new \DateTime() ) / 3600;
+	}
+
+
+	private function getStoredHour( int $hour, string $minutes ): string {
+		return '0' . ( $this->getHourInTimeZone( $hour ) ) . $minutes . 'h';
+	}
+
+	private function getMockDatabaseData(): array {
+		return [
+			[
+				'referer_name' => self::TEST_REFERER,
+				'referer_keyword' => 'testkeyword01',
+				'timestamp_floored' => '2018-05-30 0' . self::TEST_HOUR . ':' . self::TEST_MINUTES . ':00',
+				'idsite' => '1',
+				'numVisitors' => 5 ],
+			[
+				'referer_name' => self::TEST_REFERER,
+				'referer_keyword' => 'testkeyword02',
+				'timestamp_floored' => '2018-05-30 0' . self::TEST_HOUR . ':' . self::TEST_MINUTES . ':00',
+				'idsite' => '1',
+				'numVisitors' => 15 ],
+			[
+				'referer_name' => self::TEST_REFERER,
+				'referer_keyword' => 'testkeyword03',
+				'timestamp_floored' => '2018-05-30 0' . self::TEST_HOUR . ':' . self::TEST_MINUTES . ':00',
+				'idsite' => '1',
+				'numVisitors' => 20 ],
+			[
+				'referer_name' => self::TEST_REFERER,
+				'referer_keyword' => 'testkeyword04',
+				'timestamp_floored' => '2018-05-30 0' . self::TEST_HOUR . ':' . self::TEST_MINUTES_ALT . ':00',
+				'idsite' => '1',
+				'numVisitors' => 33 ]
```

Comment:
For testing the functionality more thoroughly, I think you should add:
- another entry for `testkeyword04` to check if the keywords are summed up.
- an entry with a different `referer_name` to check that campaigns are counted individually.

Reply:
>another entry for testkeyword04 to check if the keywords are summed up.

They are not actually summed up, the database query returns only one keyword for every 15 minute block, so there can never be the same keyword twice for a given time and referrer. Not sure if this 'not behavior' should be reflected in tests?!

>an entry with a different referer_name to check that campaigns are counted individually.
Added that!